nova_vm/src/ecmascript/builtins/indexed_collections/array_objects/array_prototype.rs

```diff
@@ -3034,13 +3034,57 @@ impl ArrayPrototype {
         Ok(a.into_value())
     }
 
+    /// ### [23.1.3.34 Array.prototype.toSorted ( comparator )](https://tc39.es/ecma262/#sec-array.prototype.tosorted)
     fn to_sorted(
-        _agent: &mut Agent,
-        _gc: GcScope<'_, '_>,
-        _this_value: Value,
-        _: ArgumentsList,
+        agent: &mut Agent,
+        mut gc: GcScope<'_, '_>,
+        this_value: Value,
+        args: ArgumentsList,
     ) -> JsResult<Value> {
-        todo!();
+        let comparator = args.get(0);
+        // 1. If comparator is not undefined and IsCallable(comparator) is false, throw a TypeError exception.
+        let comparator = if comparator.is_undefined() {
+            None
+        } else if let Some(comparator) = is_callable(comparator) {
+            Some(comparator)
+        } else {
+            return Err(agent.throw_exception_with_static_message(
+                gc.nogc(),
+                ExceptionType::TypeError,
+                "The comparison function must be either a function or undefined",
+            ));
+        };
+        // 2. Let o be ? ToObject(this value).
+        let o = to_object(agent, gc.nogc(), this_value)?;
+        // 3. Let len be ? LengthOfArrayLike(obj).
+        let len = usize::try_from(length_of_array_like(agent, gc.reborrow(), o)?).unwrap();
+        // 4. Let A be ? ArrayCreate(len).
+        let a = array_create(agent, gc.nogc(), len as usize, len as usize, None)?;
+        // 5. Let SortCompare be a new Abstract Closure with parameters (x, y)
+        //     that captures comparator and performs the following steps when
+        //     called:
+        //       a. Return ? CompareArrayElements(x, y, comparator).
+        // 6. Let sortedList be ? SortIndexedProperties(O, len, SortCompare, read-through-holes).
+        let sorted_list: Vec<Value> =
+            sort_indexed_properties::<false, false>(agent, gc.reborrow(), o, len, comparator)?;
+        // 7. Let j be 0.
+        let mut j = 0;
+        // 8. Repeat, while j < len,
+        while j < len {
```

Comment:
**suggestion:** You can add a fast path if `a` is `Value::Array` where you take Array slice `as_mut` and copy from the `sorted_list` directly.

Reply:
`array_create` always returns an `Array` tho, does that mean I should take the fast path every time?

Reply:
Yup :)

And this is actually to be expected: The spec uses `!` (`unwrap()` for us) because it knows exactly what path the code will take. It knows that the data property doesn't already exist and that the array being created isn't a Proxy. Because this is `CreateDataProperty` which is equivalent to `Object.defineProperty` it also means that no setters are called. Effectively, no JavaScript code can run here.

Thus, we too can simply take the path of "least resistance" and directly do what the code would end up doing anyway, which is to write data into the Array backing store.

Reply:
Doesn't that mean everything after 7. is unreachable?

Reply:
Well, not "unreachable" in the sense that it shouldn't happen but yeah, it's "unreachable" in the sense that we don't need to do those things concretely as written by the spec.

If you want to verify that what we're doing here is correct, the spec path is basically this:
1. Loop over j = 0..len
2. Get j'th Value from sorted Values
3. Call https://tc39.es/ecma262/#sec-createdatapropertyorthrow with that Value
4. Call https://tc39.es/ecma262/#sec-createdataproperty
5. Call https://tc39.es/ecma262/#sec-array-exotic-objects-defineownproperty-p-desc
    - Here we take path 2.h.
6. Call https://tc39.es/ecma262/#sec-ordinarydefineownproperty
    - Here we take path 2.d.i with `[[Writable]]: true, [[Enumerable]]: true, [[Configurable]]: true` and return `true` from 2.e.

And since in Nova we avoid storing `WEC` descriptors when all the values are true, the only data we store in OrdinaryDefineOwnProperty's step 2.d.i is the value itself, and we store that into the Array's backing store.

So, in the end we do:
1. Loop over j = 0..len
2. Get j'th Value from sorted Values
3. Store that Value to Array's backing store.

Reply:
Alright, thanks for the explanation!